app/src/main/kotlin/org/gradle/tools/trace/app/BuildOperationLog.kt

```diff
@@ -0,0 +1,97 @@
+package org.gradle.tools.trace.app
+
+interface BuildOperationLog {
+    val id: Long
+}
+
+class BuildOperationStart(
+    override val id: Long,
+    val displayName: String,
+    val startTime: Long,
+    details: Map<String, *>? = null,
+    val detailsClassName: String? = null,
+    val parentId: Long? = null
+) : BuildOperationLog {
+    val details: Map<String, Any?>? = details?.toMap()
+
+    override fun toString(): String {
+        return "BuildOperationStart{$id->$displayName}"
+    }
+}
+
+class BuildOperationFinish(
+    override val id: Long,
+    val endTime: Long,
+    result: Map<String, *>? = null,
+    val resultClassName: String? = null,
+    val failure: String? = null,
+) : BuildOperationLog {
+    val result: Map<String, Any?>? = result?.toMap()
+
+    override fun toString(): String {
+        return "BuildOperationFinish{$id}"
+    }
+}
+
+class BuildOperationProgress(
+    override val id: Long,
+    val time: Long,
+    details: Map<String, Any?>?,
+    val detailsClassName: String?
+) : BuildOperationLog {
+    val details: Map<String, Any?>? = details?.toMap()
+
+    override fun toString(): String {
+        return "Progress{details=$details, detailsClassName='$detailsClassName'}"
+    }
+}
+
+data class BuildOperationLogs(
+    val logs: List<BuildOperationLog>,
+    val include: Regex? = null,
+    val exclude: Regex? = null,
+)
+
+typealias PostVisit = (BuildOperationStart, BuildOperationFinish) -> Unit
+
+interface BuildOperationVisitor {
+
+    /**
+     * Visits the current build operation record.
+     *
+     * Returns a post-visit callback that will be called after all children have been visited.
+     */
+    fun visit(start: BuildOperationStart): PostVisit
+
+    companion object {
+
+        fun visitLogs(traversal: BuildOperationLogs, visitor: BuildOperationVisitor) {
+            val include = traversal.include
+            val exclude = traversal.exclude
+            val openBuildOperations = mutableMapOf<Long, Pair<BuildOperationStart, PostVisit>>()
+            fun helper(log: BuildOperationLog) {
+                when (log) {
+                    is BuildOperationStart -> {
+                        val displayName = log.displayName
+
+                        val included = include == null || displayName.matches(include) || openBuildOperations.contains(log.parentId)
+                        if (!included || (exclude != null && displayName.matches(exclude))) {
+                            return
+                        }
+                        openBuildOperations[log.id] = log to visitor.visit(log)
```

Comment:
Why are we doing this weird mapping from `start event` to `function that handles start event and finish event`?

Can't we just call `visitor.visit(openBuildOp).invoke(openBuildOp, log)` below? (and then potentially clean that up into a single method call?)

Or, do we need to call the visitor immediately once seeing the start event? 

In any case, I think if we split the `BuildOperationVisitor` to have two methods `visitStart` and `visitFinish` we won't need this mapping anymore either

Reply:
> Or, do we need to call the visitor immediately once seeing the start event?

Yes, we want to do that so we have proper nesting data. For the visitor, it is nicer to return the PostVisit method, so it doesn't need to keep internal state.